## This NB serves the purpose of concatenate the continente and pingo doce datasets. And filter the full dataset 2 times: one time for further analysis and a second time to provide consumer information. 

- Note: this NB was made in google colab.

In [27]:
import time
import json

import random
import datetime

import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np

from google.colab import drive

import chardet
import re

In [28]:
drive.mount('drive')

Drive already mounted at drive; to attempt to forcibly remount, call drive.mount("drive", force_remount=True).


### Concatenate the data sets in the intersected columns and respecting the indexes.

In [29]:
continente = pd.read_csv(r'conti.csv',  encoding='utf-8')
pingo_doce = pd.read_csv(r'pingo_doce.csv',  encoding='utf-8')

In [30]:
continente

Unnamed: 0       id          Category  \
0               0  2076239         mercearia   
1               1  7405690         mercearia   
2               2  2113579         mercearia   
3               3  2305902         mercearia   
4               4  7106089         mercearia   
...           ...      ...               ...   
11260       12246  4374082  refeicoes-faceis   
11261       12247  6561375  refeicoes-faceis   
11262       12248  7551410  refeicoes-faceis   
11263       12249  7600279  refeicoes-faceis   
11264       12250  7378676  refeicoes-faceis   

                                       Name  Amount Unit  List_Price  \
0                            Atum em Azeite    0.12   KG        1.22   
1             Cápsulas de Café Buondi Int 8   64.00   UN       18.99   
2                    Atum Filetes em Azeite    0.12   KG        2.53   
3                              Arroz Agulha    1.00   KG        1.59   
4      Batata Frita Gourmet Lisa Corte Fino    0.15   KG        2.69   
...                                     ...     ...  ...         ...   
11260               Massa Fusilli com Pesto    0.15   KG        1.49   
11261        Hambúrguer Sementes sem Glúten    0.16   KG        3.39   
11262          Tagliatelle Fresco Biológico    0.25   UN        2.19   
11263                     Pastel com Salmão    1.00   KG        1.19   
11264                      Empada de Cozido    2.00   KG        0.99   

      Price_per_unit_of_measure Currency  
0                         14.19        €  
1                          0.30        €  
2                         32.44        €  
3                          1.59        €  
4                         17.93        €  
...                         ...      ...  
11260                      9.93        €  
11261                     21.19        €  
11262                      2.19        €  
11263                     14.00        €  
11264                     12.38        €  

[11265 rows x 9 columns]

In [31]:
pingo_doce

Unnamed: 0         Time scraped     Category  \
0              0  2023-05-12 18:26:22    mercearia   
1              1  2023-05-12 18:26:22    mercearia   
2              2  2023-05-12 18:26:22    mercearia   
3              3  2023-05-12 18:26:22    mercearia   
4              4  2023-05-12 18:26:22    mercearia   
...          ...                  ...          ...   
5596        6179  2023-05-12 16:32:53  frigorífico   
5597        6180  2023-05-12 16:32:53  frigorífico   
5598        6181  2023-05-12 16:32:53  frigorífico   
5599        6182  2023-05-12 16:32:53  frigorífico   
5600        6183  2023-05-12 16:32:53  frigorífico   

                                                   Name  Amount Units  \
0                     Doce de Abóbora Sabores do Campo    0.310    KG   
1                   Sal em Saco Rui Simeão Marinho DOP    1.500    KG   
2                      Doce de Tomate Sabores do Campo    0.310    KG   
3                       Snack Hóstias de Camarão Zanuy    0.050    KG   
4                           Cacau Cru em Pepitas Seara    0.100    KG   
...                                                 ...     ...   ...   
5596    Iogurte Soja sem Açúcar Frutos Vermelhos Alpro    0.135    KG   
5597   Iogurte Líquido sem Lactose de Mirtilo e Chá ...   0.624    KG   
5598                Iogurte Skyr de Soja Natural Alpro    0.400    KG   
5599                Iogurte Skyr de Soja Morango Alpro    0.400    KG   
5600   Iogurte Sólido Magro Coco Manga e Pêssego Mim...   0.960    KG   

      List_Price  Price_per_unit_of_measure Currency  
0           1.99                       6.42        €  
1           1.37                       0.91        €  
2           1.99                       6.42        €  
3           1.79                      35.80        €  
4           3.99                      39.90        €  
...          ...                        ...      ...  
5596        1.16                       8.59        €  
5597        3.10                      19.87        €  
5598        2.82                       7.05        €  
5599        2.82                       7.05        €  
5600        2.34                       2.44        €  

[5601 rows x 9 columns]

In [32]:
pingo_doce.columns

Index(['Unnamed: 0', 'Time scraped', 'Category', 'Name', 'Amount', 'Units',
       'List_Price', 'Price_per_unit_of_measure', 'Currency'],
      dtype='object')

In [33]:
continente.columns

Index(['Unnamed: 0', 'id', 'Category', 'Name', 'Amount', 'Unit', 'List_Price',
       'Price_per_unit_of_measure', 'Currency'],
      dtype='object')

In [34]:
## add column Store to dataset
continente['Store'] = 'continete'
pingo_doce['Store'] = 'pingo_doce'

In [35]:
## rename column
continente = continente.rename(columns={"Unnamed: 0": "Old_index",})

In [36]:
## increment value of Index
pingo_doce['Old_index'] = pingo_doce['Unnamed: 0'] + len(continente)

In [37]:
intersec_columns = []

for column in continente.columns:
  if column in pingo_doce.columns:
    intersec_columns.append(column)

intersec_columns

['Old_index',
 'Category',
 'Name',
 'Amount',
 'List_Price',
 'Price_per_unit_of_measure',
 'Currency',
 'Store']

In [38]:
continente_concat = continente[intersec_columns]

pingo_doce_concat = pingo_doce[intersec_columns]

frames = [continente_concat, pingo_doce_concat]

df_concat = pd.concat(frames)

df_concat = df_concat.reset_index()
df_concat = df_concat.drop(['index'], axis=1)

df_concat

Old_index     Category  \
0              0    mercearia   
1              1    mercearia   
2              2    mercearia   
3              3    mercearia   
4              4    mercearia   
...          ...          ...   
16861      17444  frigorífico   
16862      17445  frigorífico   
16863      17446  frigorífico   
16864      17447  frigorífico   
16865      17448  frigorífico   

                                                    Name  Amount  List_Price  \
0                                         Atum em Azeite   0.120        1.22   
1                          Cápsulas de Café Buondi Int 8  64.000       18.99   
2                                 Atum Filetes em Azeite   0.120        2.53   
3                                           Arroz Agulha   1.000        1.59   
4                   Batata Frita Gourmet Lisa Corte Fino   0.150        2.69   
...                                                  ...     ...         ...   
16861    Iogurte Soja sem Açúcar Frutos Vermelhos Alpro    0.135        1.16   
16862   Iogurte Líquido sem Lactose de Mirtilo e Chá ...   0.624        3.10   
16863                Iogurte Skyr de Soja Natural Alpro    0.400        2.82   
16864                Iogurte Skyr de Soja Morango Alpro    0.400        2.82   
16865   Iogurte Sólido Magro Coco Manga e Pêssego Mim...   0.960        2.34   

      Price_per_unit_of_measure Currency       Store  
0                         14.19        €   continete  
1                          0.30        €   continete  
2                         32.44        €   continete  
3                          1.59        €   continete  
4                         17.93        €   continete  
...                         ...      ...         ...  
16861                      8.59        €  pingo_doce  
16862                     19.87        €  pingo_doce  
16863                      7.05        €  pingo_doce  
16864                      7.05        €  pingo_doce  
16865                      2.44        €  pingo_doce  

[16866 rows x 8 columns]

Standerdization of categories

In [39]:

category_mapping = {
    'mercearia': ['mercearia', 'bio-e-escolhas-alimentares'],
    'padaria_pastelaria': ['padaria_pastelaria', 'padaria-e-pastelaria'],
    'frutas_legumes': ['frutas_legumes', 'frutas-e-legumes'],
    'peixaria_talho': ['talho', 'peixaria', 'peixaria-e-talho'],
    'leite_ovos': ['leite_ovos_natas', 'laticinios-e-ovos'],
    'congelados': ['congelados'],
    'frigorifico': ['take-away', 'frigorífico', 'refeicoes-faceis', 'charcutaria-e-queijos']
}

conditions = [
    df_concat['Category'].isin(category_mapping['mercearia']),
    df_concat['Category'].isin(category_mapping['padaria_pastelaria']),
    df_concat['Category'].isin(category_mapping['frutas_legumes']),
    df_concat['Category'].isin(category_mapping['peixaria_talho']),
    df_concat['Category'].isin(category_mapping['leite_ovos']),
    df_concat['Category'].isin(category_mapping['congelados']),
    df_concat['Category'].isin(category_mapping['frigorifico'])
]

choices = [
    'mercearia',
    'padaria_pastelaria',
    'frutas_legumes',
    'peixaria_talho',
    'leite_ovos',
    'congelados',
    'frigorifico'
]

df_concat['Standardized_Category'] = np.select(conditions, choices, default=df_concat['Category'])

In [40]:
category_mapping = {
    'MERC': ['mercearia', 'bio-e-escolhas-alimentares'],
    'PAPA': ['padaria_pastelaria', 'padaria-e-pastelaria'],
    'FRLE': ['frutas_legumes', 'frutas-e-legumes'],
    'PETA': ['talho', 'peixaria', 'peixaria-e-talho'],
    'LEOV': ['leite_ovos_natas', 'laticinios-e-ovos'],
    'CONGE': ['congelados'],
    'FRIGO': ['take-away', 'frigorífico', 'refeicoes-faceis', 'charcutaria-e-queijos']
}

conditions = [
    df_concat['Category'].isin(category_mapping['MERC']),
    df_concat['Category'].isin(category_mapping['PAPA']),
    df_concat['Category'].isin(category_mapping['FRLE']),
    df_concat['Category'].isin(category_mapping['PETA']),
    df_concat['Category'].isin(category_mapping['LEOV']),
    df_concat['Category'].isin(category_mapping['CONGE']),
    df_concat['Category'].isin(category_mapping['FRIGO'])
]

choices = [
    'MERC',
    'PAPA',
    'FRLE',
    'PETA',
    'LEOV',
    'CONGE',
    'FRIGO'
]

df_concat['Abbr_Category'] = np.select(conditions, choices, default=df_concat['Category'])

In [41]:
df_concat.drop(['Category'], axis = 1, inplace = True)

### First filter. We filtered the first time with 1 token of each foodstuff (alimento) as one can see through the "basic_cabaz" list!

In [42]:
## selecting just the products with name of basic cabaz

In [43]:
#basic_cabaz = ['Leite', 'Queijo', 'Pão', 'Arroz', 'Massa', 'Cereal', 'Tosta', 'Bolacha', 'Feijão', 'Grão-de-bico', 'Ervilhas', 
   #            'Frango', 'Pescada', 'Atum', 'Sardinha', 'Cavala', 'Tomate', 'Vegetais', 'Brocolos', 'Espinafre', 'Cenoura', 'Alho', 'Azeite', 'Creme Vegetal', 'Marmelada']

basic_cabaz = [
    'Leite', 'Queijo', 'Arroz', 'Massa', 'Cereal', 'Tosta', 'Bolacha Maria', 
    'Feijão', 'Grão de bico', 'Ervilha', 'Frango', 'Pescada', 'Atum', 'Sardinha',
    'Cavala', 'Tomate', 'Vegetais', 'Brócolos', 'Feijão Verde', 'Espinafre', 'Cenoura',
    'Alho Francês', 'Azeite', 'Creme Vegetal', 'Marmelada'
]

In [44]:
# Create an empty DataFrame
all_filtered_data = pd.DataFrame()

for index, basic_product in enumerate(basic_cabaz):
    pattern = r'{}'.format(basic_product)
    filtered = df_concat[df_concat['Name'].str.contains(pattern, flags=re.IGNORECASE)].copy()
    filtered.loc[:, 'Cabaz_product'] = index

    # Remove the values already present in all_filtered_data
    if len(all_filtered_data) > 0:
        filtered_df = filtered[~filtered['Old_index'].isin(all_filtered_data['Old_index'])]
    else:
        filtered_df = filtered

    all_filtered_data = pd.concat([all_filtered_data, filtered_df])

df_all_filtered_data = all_filtered_data.sort_index() 

In [45]:
df_all_filtered_data.count()

Old_index                    3629
Name                         3629
Amount                       3629
List_Price                   3629
Price_per_unit_of_measure    3629
Currency                     3629
Store                        3629
Standardized_Category        3629
Abbr_Category                3629
Cabaz_product                3629
dtype: int64

In [46]:
df_all_filtered_data['Standardized_Category'].value_counts()

mercearia             2000
frigorifico            860
leite_ovos             226
congelados             206
peixaria_talho         154
padaria_pastelaria     104
frutas_legumes          79
Name: Standardized_Category, dtype: int64

In [47]:
df_all_filtered_data

Old_index                                               Name  Amount  \
0              0                                     Atum em Azeite   0.120   
2              2                             Atum Filetes em Azeite   0.120   
3              3                                       Arroz Agulha   1.000   
5              5                                   Massa Esparguete   0.500   
7              7                                    Atum ao Natural   0.120   
...          ...                                                ...     ...   
16846      17392                      Queijo Fundido A Vaca que Ri    0.125   
16847      17393                Queijo Fundido Light A Vaca que Ri    0.125   
16848      17394   Smoothie+ Leite de Coco e Frutos Vermelhos Pi...   0.250   
16849      17395           Queijo de Ovelha Grande Seia do Tavares    0.990   
16857      17403                                   Massa Filó Zimi    0.300   

       List_Price Price_per_unit_of_measure Currency       Store  \
0            1.22                     14.19        €   continete   
2            2.53                     32.44        €   continete   
3            1.59                      1.59        €   continete   
5            1.16                      2.32        €   continete   
7            1.12                     13.33        €   continete   
...           ...                       ...      ...         ...   
16846        1.59                     12.72        €  pingo_doce   
16847        1.59                     12.72        €  pingo_doce   
16848        0.99                      3.96        €  pingo_doce   
16849       16.30                      16.3        €  pingo_doce   
16857        2.79                       9.3        €  pingo_doce   

      Standardized_Category Abbr_Category  Cabaz_product  
0                 mercearia          MERC             12  
2                 mercearia          MERC             12  
3                 mercearia          MERC              2  
5                 mercearia          MERC              3  
7                 mercearia          MERC             12  
...                     ...           ...            ...  
16846           frigorifico         FRIGO              1  
16847           frigorifico         FRIGO              1  
16848           frigorifico         FRIGO              0  
16849           frigorifico         FRIGO              1  
16857           frigorifico         FRIGO              3  

[3629 rows x 10 columns]

In [48]:
df_all_filtered_data.describe()

Old_index       Amount   List_Price  Cabaz_product
count   3629.000000  3629.000000  3629.000000    3629.000000
mean    8906.814274     0.769850     3.302824       6.281345
std     5372.513097     3.790185     3.342309       6.926408
min        0.000000     0.001080     0.150000       0.000000
25%     3878.000000     0.150000     1.570000       1.000000
50%     9775.000000     0.300000     2.350000       3.000000
75%    12927.000000     0.500000     3.760000      11.000000
max    17403.000000   186.300000    35.470000      24.000000

In [49]:
df_all_filtered_data.to_csv('data_cabaz_prducts.csv')

## Second filter. Here we filter the "df_all_filtered_data" dataset that contains 3629 rows × 9 columns using rapidfuzz. Now we use all tokens of the foodstuff (alimentos) to improve the accuracy of the filter.  


Reducing the df_all_filtered_data with the rapidfuzz library

In [50]:
items_cabaz = [
    'Leite Meio Gordo', 'Queijo Meio Gordo', 'Arroz', 'Massa', 'Cereais', 'Tostas', 'Bolacha Maria',
    'Feijão lata', 'Grão-de-bico lata', 'Ervilhas lata', 'Frango inteiro congelado',
    'Pescada congelada', 'Atum lata', 'Sardinha', 'Cavala lata', 'Tomate pelado',
    'Mistura de vegetais para sopa', 'Brócolos', 'Feijão Verde', 'Espinafres', 'Cenoura',
    'Alho Francês', 'Azeite', 'Creme Vegetal', 'Marmelada'
]

In [51]:
df_re = df_all_filtered_data.copy()

In [52]:
!pip install rapidfuzz

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [54]:
# The code matches each product name of the full dataset with each of the name of the items_cabaz list and "build" a new data frame with the matched products. The WRatio was the scorer that present better results and the limit 15 is a reasonable number after testing.

from rapidfuzz import fuzz, process, utils

matches_data = pd.DataFrame(columns = df_re.columns)

for i in items_cabaz: 
  matches = matches = process.extract(i, df_re['Name'], scorer = fuzz.WRatio, limit = 15, processor = utils.default_process)
  for matched in matches:
    row = df_re.loc[df_re['Name'] == matched[0]]
    matches_data = pd.concat([matches_data, row])
  
matches_data.reset_index(drop = True, inplace = True)

In [55]:
matches_data

Old_index                  Name  Amount  List_Price  \
0         8530  Leite UHT Meio Gordo    3.00        1.11   
1         8547  Leite UHT Meio Gordo    3.00        0.89   
2         8809  Leite UHT Meio Gordo    3.00        1.11   
3         8915  Leite UHT Meio Gordo    8.00        6.40   
4         9248  Leite UHT Meio Gordo    0.25        0.92   
...        ...                   ...     ...         ...   
1231      3166      Marmelada Nature    0.25        3.99   
1232      7776      Marmelada Nature    0.25        3.99   
1233      4211    Marmelada em Cubos    0.25        3.89   
1234      3148       Marmelada Light    0.40        3.49   
1235      6929       Marmelada Light    0.40        3.49   

     Price_per_unit_of_measure Currency      Store Standardized_Category  \
0                         1.85        €  continete            leite_ovos   
1                         1.48        €  continete            leite_ovos   
2                         1.85        €  continete            leite_ovos   
3                         0.80        €  continete            leite_ovos   
4                         3.68        €  continete            leite_ovos   
...                        ...      ...        ...                   ...   
1231                     15.96        €  continete             mercearia   
1232                     15.96        €  continete             mercearia   
1233                     15.56        €  continete             mercearia   
1234                      8.73        €  continete             mercearia   
1235                      8.73        €  continete             mercearia   

     Abbr_Category Cabaz_product  
0             LEOV             0  
1             LEOV             0  
2             LEOV             0  
3             LEOV             0  
4             LEOV             0  
...            ...           ...  
1231          MERC            24  
1232          MERC            24  
1233          MERC            24  
1234          MERC            24  
1235          MERC            24  

[1236 rows x 10 columns]

In [56]:
matches_data['Store'].value_counts()

continete     1202
pingo_doce      34
Name: Store, dtype: int64

## Testing Zone

In [58]:
from rapidfuzz import fuzz, process, utils

for i in items_cabaz: 
  matches = process.extract(i, df_re['Name'], scorer = fuzz.WRatio, limit = 15, processor = utils.default_process)
  print(f"Matches for {i}: {matches} \n")

Matches for Leite Meio Gordo: [('Leite UHT Meio Gordo', 95.0, 8041), ('Leite UHT Meio Gordo', 95.0, 8056), ('Leite UHT Meio Gordo', 95.0, 8301), ('Leite UHT Meio Gordo', 95.0, 8399), ('Leite UHT Meio Gordo', 95.0, 8669), ('Leite Meio Gordo Fresco', 95.0, 8883), (' Leite Meio Gordo Vigor ', 95.0, 16104), ('Leite Meio Gordo sem Lactose', 90.0, 5798), ('Leite Meio Gordo Infantil sem Lactose + 3 Anos', 90.0, 7499), ('Leite Meio Gordo sem Lactose', 90.0, 8057), ('Leite Meio Gordo Infantil sem Lactose + 3 Anos', 90.0, 8808), ('Tablete de Chocolate de Leite Extrafino sem Glúten', 85.5, 9), ('Leite em Pó Confort 2 +6M sem Glúten', 85.5, 16), ('Bolachas Cobertas Chocolate de Leite', 85.5, 182), ('Bolachas Argolinhas com Cobertura Chocolate de Leite', 85.5, 320)] 

Matches for Queijo Meio Gordo: [('Queijo Meio Gordo', 100.0, 9103), ('Queijo Meio Gordo', 100.0, 9448), ('Queijo Fresco Meio Gordo', 95.0, 9101), ('Queijo Fresco Meio Gordo', 95.0, 9107), ('Queijo Meio Gordo sem Lactose', 90.0, 9346),

In [59]:
x = pd.DataFrame(matches)
x

0      1     2
0                 Marmelada  100.0  1236
1                 Marmelada  100.0  1388
2                 Marmelada  100.0  2932
3                 Marmelada  100.0  3133
4                 Marmelada  100.0  3265
5                 Marmelada  100.0  4455
6                 Marmelada  100.0  6915
7                 Marmelada  100.0  7217
8   Marmelada Pack Poupança   90.0   762
9           Marmelada Extra   90.0  1493
10          Marmelada Light   90.0  3114
11        Taça de Marmelada   90.0  3120
12         Marmelada Nature   90.0  3132
13       Marmelada em Cubos   90.0  4168
14          Marmelada Light   90.0  6562

## to remove?

In [ ]:
import pandas as pd
data = pd.read_excel('/Users/andrepedrooliveira/Documents/P4DS/Project/1200_cabaz_products_final.xlsx')
data.head()

In [ ]:
currency_to_clean = data['Currency']

In [ ]:
currency_cleaned = []

for i in currency_to_clean:
    x = i.replace('€', '€')
    currency_cleaned.append(x)

In [ ]:
data.drop(['Currency'], axis = 1, inplace = True)

In [ ]:
currency = currency_cleaned

In [ ]:
file_path = "/Users/andrepedrooliveira/Documents/P4DS/Project/data_info_final.csv"

data.to_csv(file_path, index=False)